test/run-functional-tests

```diff
@@ -0,0 +1,430 @@
+		printf "%3s    %-40s %-60s\n" $index $f "$desc"
+		index=$(($index+1))
+	done
+	printf "%3s    %-40s  %-60s\n" $index "regression test" "Original regression test"
+}
+
+
+usage_and_exit() {
+	prog=`basename $0`
+	cat <<END
+Usage: $prog [OPTIONS]|[TESTCASE INDEX]
+$prog is a tool for developers to setup the cluster in containers to run functional tests.
+The container image is based on Tumbleweed with preinstalled packages of the cluster stack include pacemaker/corosync/crmsh and many others.
+Users can make the code change under crmsh.git including test cases. This tool will pick up the code change and "make install" to all running containers.
+
+OPTIONS:
+  -h, --help		Show this help message and exit
+  -l            	List existing functional test cases and exit
+  -n NUM        	Only setup a cluster with NUM nodes(containers)
+  -x			Don't config corosync on containers(with -n option)
+  -d            	Cleanup the cluster containers
+
+EXAMPLES:
+To launch 2 nodes with the running cluster with the very basic corosync.conf
+# crmsh.git/test/run-functional-tests -n 2
+
+To launch 2 nodes without the cluster stack running to play with "crm cluster init/join"
+# crmsh.git/run-functional-tests -n 2 -x
+
+To list the existing test cases. Users could add his own new test cases.
+# crmsh.git/test/run-functional-tests -l
+
+To run a single or a number of functional test cases
+# crmsh.git/test/run-functional-tests 1
+# crmsh.git/test/run-functional-tests 1 2 3
+
+To clean up the all containers which are generated by this tool
+# crmsh.git/test/run-functional-tests -d
+END
+	exit 1
+}
+
+
+docker_exec() {
+	name=$1
+	cmd=$2
+	docker exec -t $name /bin/sh -c "$cmd"
+}
+
+
+pre_make_on_local() {
+        project_path=$1
+        cd $project_path
+        for cmd in autoconf automake;do
+                which $cmd &> /dev/null || fatal "required $cmd!"
+        done
+        ./autogen.sh &> /dev/null && ./configure --prefix /usr &> /dev/null
+}
+
+
+deploy_ha_node() {
+	node_name=$1
+	project_path=$(dirname $(dirname `realpath $0`))
+	docker_options="-d --name $node_name -h $node_name --privileged --shm-size 1g -v $project_path:/app"
+        build_cmd="cd /app;make install && make install-crmconfDATA prefix="
+
+        pre_make_on_local $project_path || fatal "build failed!"
+	info "Deploying \"$node_name\"..."
+	docker run --restart always $docker_options $DOCKER_IMAGE &> /dev/null
+	for network in ${HA_NETWORK_ARRAY[@]};do
+		docker network connect $network $node_name
+	done
+
+	if [ "$node_name" == "node-without-ssh" ];then
+		docker_exec $node_name "systemctl stop sshd.service"
+	fi
+	if [ "$node_name" != "qnetd-node" ];then
+		docker_exec $node_name "rpm -e corosync-qnetd"
+	fi
+	docker_exec $node_name "rm -rf /run/nologin"
+	docker_exec $node_name "echo 'StrictHostKeyChecking no' >> /etc/ssh/ssh_config"
+	info "Building crmsh codes on \"$node_name\"..."
+	docker_exec $node_name "$build_cmd" 1> /dev/null || \
```

Comment:
I can't help seeing some magic, when run `$ ./run-functional-tests -n 2` it shoots the "Building failed!", even that there was no error in the `docker_exec $node_name "$build_cmd"` command. And when creating only 1 node cluster `$ ./run-functional-tests -n 1` it works like charm. @liangxin1300 have you encountered such behavior?